drivers/sdhc/intel_emmc_host.c

```diff
@@ -0,0 +1,1342 @@
+			      EMMC_HOST_XFER_COMPLETE |
+				      ERR_INTR_STATUS_EVENT(EMMC_HOST_DMA_TXFR_ERR),
+			      false, wait_time);
+
+	if (events & EMMC_HOST_XFER_COMPLETE) {
+		ret = 0;
+	} else if (events & ERR_INTR_STATUS_EVENT(0xFFFF)) {
+		LOG_ERR("wait for xfer complete error: %x", events);
+		ret = -EIO;
+	} else {
+		LOG_ERR("wait for xfer complete timeout");
+		ret = -EAGAIN;
+	}
+
+	return ret;
+}
+
+static int wait_xfr_poll_complete(const struct device *dev, uint32_t time_out)
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+	int ret = -EAGAIN;
+	int32_t retry = time_out;
+
+	LOG_DBG("");
+
+	while (retry > 0) {
+		if (regs->normal_int_stat & EMMC_HOST_XFER_COMPLETE) {
+			regs->normal_int_stat |= EMMC_HOST_XFER_COMPLETE;
+			ret = 0;
+			break;
+		}
+
+		k_busy_wait(EMMC_HOST_MSEC_DELAY);
+		retry--;
+	}
+
+	return ret;
+}
+
+static int wait_xfr_complete(const struct device *dev, uint32_t time_out)
+{
+	int ret;
+
+	if (IS_ENABLED(CONFIG_INTEL_EMMC_HOST_INTR)) {
+		ret = wait_xfr_intr_complete(dev, time_out);
+	} else {
+		ret = wait_xfr_poll_complete(dev, time_out);
+	}
+	return ret;
+}
+
+static enum emmc_response_type emmc_decode_resp_type(enum sd_rsp_type type)
+{
+	enum emmc_response_type resp_type;
+
+	switch (type & 0xF) {
+	case SD_RSP_TYPE_NONE:
+		resp_type = EMMC_HOST_RESP_NONE;
+		break;
+	case SD_RSP_TYPE_R1:
+	case SD_RSP_TYPE_R3:
+	case SD_RSP_TYPE_R4:
+	case SD_RSP_TYPE_R5:
+		resp_type = EMMC_HOST_RESP_LEN_48;
+		break;
+	case SD_RSP_TYPE_R1b:
+		resp_type = EMMC_HOST_RESP_LEN_48B;
+		break;
+	case SD_RSP_TYPE_R2:
+		resp_type = EMMC_HOST_RESP_LEN_136;
+		break;
+
+	case SD_RSP_TYPE_R5b:
+	case SD_RSP_TYPE_R6:
+	case SD_RSP_TYPE_R7:
+	default:
+		resp_type = EMMC_HOST_INVAL_HOST_RESP_LEN;
+	}
+
+	return resp_type;
+}
+
+static void update_cmd_response(const struct device *dev, struct sdhc_command *sdhc_cmd)
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+	uint32_t resp0, resp1, resp2, resp3;
+
+	if (sdhc_cmd->response_type == SD_RSP_TYPE_NONE) {
+		return;
+	}
+
+	resp0 = regs->resp_01;
+
+	if (sdhc_cmd->response_type == SD_RSP_TYPE_R2) {
+		resp1 = regs->resp_2 | (regs->resp_3 << 16u);
+		resp2 = regs->resp_4 | (regs->resp_5 << 16u);
+		resp3 = regs->resp_6 | (regs->resp_7 << 16u);
+
+		LOG_DBG("cmd resp: %x %x %x %x", resp0, resp1, resp2, resp3);
+
+		sdhc_cmd->response[0u] = resp3;
+		sdhc_cmd->response[1U] = resp2;
+		sdhc_cmd->response[2U] = resp1;
+		sdhc_cmd->response[3U] = resp0;
+	} else {
+		LOG_DBG("cmd resp: %x", resp0);
+		sdhc_cmd->response[0u] = resp0;
+	}
+}
+
+static int32_t emmc_host_send_cmd(const struct device *dev, const struct emmc_cmd_config *config)
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+	struct emmc_data *emmc = dev->data;
+	struct sdhc_command *sdhc_cmd = config->sdhc_cmd;
+	enum emmc_response_type resp_type = emmc_decode_resp_type(sdhc_cmd->response_type);
+	uint16_t cmd_reg;
+	int ret;
```

Comment:
Here you're also mixing `int32_t` in the function prototype, but using `int ret` inside the function. There are several more cases of this in the driver. I suppose every place where you use the "0 on success or negative POSIX errno on failure" pattern you should use `int` as the type.